### ISSUE TITLE
DNS resolution fails with root search domain

### ISSUE THREAD

**Issue by michaelkuhn** [erlang/otp#4838]
**Describe the bug**
After upgrading my server to Fedora 34, I started getting errors such as the following in ejabberd. This problem has already been reported in ejabberd (https://github.com/processone/ejabberd/issues/3438) but was deemed to be a problem in Erlang itself.

The problem stems from the fact that recent versions of systemd-resolved started putting `search .` into `/etc/resolv.conf` by default, which apparently causes problems within Erlang. After changing it to `search localdomain`, the problem went away.

```
2021-05-16 13:23:14.171491+02:00 [error] <0.396.0>@supervisor:do_restart/3:722 SUPERVISOR REPORT:
    supervisor: {local,ejabberd_s2s_out_sup}
    errorContext: child_terminated
    reason: {function_clause,
                [{inet_dns,encode_labels,
                     [<<...>>,
                      {5,
                       {[<<"_xmpps-server">>,<<"_tcp">>,<<"p2">>,<<"siacs">>,
                         <<"eu">>,<<>>],
                        12,
                        {[<<"_tcp">>,<<"p2">>,<<"siacs">>,<<"eu">>,<<>>],
                         26,nil,nil},
                        {[<<"p2">>,<<"siacs">>,<<"eu">>,<<>>],
                         31,
                         {[<<"eu">>,<<>>],40,nil,nil},
                         {[<<"siacs">>,<<"eu">>,<<>>],34,nil,nil}}}},
                      43,
                      [<<>>]],
                     [{file,"inet_dns.erl"},{line,694}]},
                 {inet_dns,encode_name,4,[{file,"inet_dns.erl"},{line,675}]},
                 {inet_dns,encode_query_section,3,
                     [{file,"inet_dns.erl"},{line,269}]},
                 {inet_dns,encode,1,[{file,"inet_dns.erl"},{line,240}]},
                 {inet_res,make_query,5,[{file,"inet_res.erl"},{line,650}]},
                 {inet_res,make_query,4,[{file,"inet_res.erl"},{line,618}]},
                 {inet_res,res_query,6,[{file,"inet_res.erl"},{line,602}]},
                 {inet_res,res_getby_query,4,
                     [{file,"inet_res.erl"},{line,569}]}]}
    offender: [{pid,<0.727.0>},
               {id,undefined},
               {mfargs,{ejabberd_s2s_out,start_link,undefined}},
               {restart_type,temporary},
               {shutdown,5000},
               {child_type,worker}]
```

**To Reproduce**
1. Put `search .` into `/etc/resolv.conf`
2. Try to resolve hostname

**Expected behavior**
DNS resolution should work even with `search .` in `/etc/resolv.conf`.

**Affected versions**
23.3.2-1.fc34

**Additional context**
The problem happens with ejabberd 20.07-2.fc34 (if this is relevant).

**Comment by RaimoNiskanen** [comment]
The `.` search domain was not anticipated at all by the code, so joining the lookup name with the `.` search domain probably produced `_xmpps-server._tcp.p2.siacs.eu..`, and the subsequent code saw the `..` as an empty label and barfed.

This should fix it, diff for OTP-24.0.  I will check with the team on Monday how far back it would be reasonable to fix this...
```erlang
@@ -537,31 +537,41 @@ type_p(Type) ->
 %%
 res_getbyname(Name, Type, Timer) ->
     {EmbeddedDots, TrailingDot} = inet_parse:dots(Name),
-    Dot = if TrailingDot -> ""; true -> "." end,
-    if  TrailingDot ->
+    if
+        TrailingDot ->
 	    res_getby_query(lists:droplast(Name), Type, Timer);
 	EmbeddedDots =:= 0 ->
-	    res_getby_search(Name, Dot,
-			     inet_db:get_searchlist(),
+	    res_getby_search(Name, inet_db:get_searchlist(),
 			     nxdomain, Type, Timer);
 	true ->
 	    case res_getby_query(Name, Type, Timer) of
 		{error,_Reason}=Error ->
-		    res_getby_search(Name, Dot,
-				     inet_db:get_searchlist(),
+		    res_getby_search(Name, inet_db:get_searchlist(),
 				     Error, Type, Timer);
 		Other -> Other
 	    end
     end.
 
-res_getby_search(Name, Dot, [Dom | Ds], _Reason, Type, Timer) ->
-    case res_getby_query(Name++Dot++Dom, Type, Timer,
+res_getby_search(Name, [Dom | Ds], _Reason, Type, Timer) ->
+    QueryName =
+        %% Join Name and Dom with a single dot.
+        %% Allow Dom to be "." or "", but not to lead with ".".
+        %% Do not allow Name to be "".
+        if
+            Name =/= "" andalso (Dom =:= "." orelse Dom =:= "") ->
+                Name;
+            Name =/= "" andalso hd(Dom) =/= $. ->
+                Name++"."++Dom;
+            true ->
+                erlang:error({if_clause, Name, Dom})
+        end,
+    case res_getby_query(QueryName, Type, Timer,
 			 inet_db:res_option(nameservers)) of
 	{ok, HEnt}         -> {ok, HEnt};
 	{error, NewReason} ->
-	    res_getby_search(Name, Dot, Ds, NewReason, Type, Timer)
+	    res_getby_search(Name, Ds, NewReason, Type, Timer)
     end;
-res_getby_search(_Name, _, [], Reason,_,_) ->
+res_getby_search(_Name, [], Reason,_,_) ->
     {error, Reason}.
 
 res_getby_query(Name, Type, Timer) ->
```

**Comment by RaimoNiskanen** [comment]
I accidentally fixed this on an unrelated topic branch, that goes in to OTP-24.0.2 to be released today or tomorrow.

If this needs to be backported to an earlier release, please let me know.  It is a bit awkward, but can be done...

**Comment by michaelkuhn** [comment]
I guess it would be nice if the fix could be ported back to 23.3.4.x (not exactly sure about the versioning scheme), which is used in Fedora 34 (see https://src.fedoraproject.org/rpms/erlang) because it breaks the out-of-the-box ejabberd installation there.

**Comment by RaimoNiskanen** [comment]
Okidokey. Backporting...  Target patch release OTP 23.3.4.3.

**Comment by RaimoNiskanen** [comment]
Rebase went smooth.  Daily tests now.  Probably merge on monday, and release in that week.

**Comment by RaimoNiskanen** [comment]
Backported and released in OTP-23.3.4.3
